src/regularization_smooth.jl

```diff
@@ -0,0 +1,109 @@
+# TODO:
+# - midpoint rule
+# x scattered/interpolation
+# x GCV
+# x L-curve
+# - bounds on λ
+# - initial guess for λ? will require mods to RegularizationTools
+# - scaled λ? will need to work out equivalency with λ² formulation, and resolve with
+#   derivative rather than difference matrix
+# - standard deviation
+# - relative weights? (not sure if worth implementing)
+# - arbitrary weighting
+
+
+const LA = LinearAlgebra
+
+# might move this to RegularizationTools.jl? JJS 12/6/21
+"""
+Order d derivative matrix for the provided t vector
+"""
+function _derivative_matrix(t::AbstractVector, d::Int)
+    N = length(t)
+    if d == 0
+        return Array{Float64}(LA.I, (N, N))
+    end
+    dt = t[d+1:end] - t[1:end-d]
+    V = LA.diagm(1 ./ dt)
+    # could use this symbol:  Dᵈ⁻¹ (D\^d\^-\^1) (but a lot of typing...)
+    Ddm1 = diff(_derivative_matrix(t, d-1), dims=1)
+    D = d * V*Ddm1
+    return D
+end
+
+### Regularization data smoothing and interpolation
+struct RegularizationSmooth{uType,tType,FT,T} <: AbstractInterpolation{FT,T}
+    u::uType
+    û::uType
+    t::tType
+    t̂::tType
+    d::Int              # derivative degree used to calculate the roughness
+    λ::AbstractFloat    # regularization parameter
+    alg::Symbol  # how to determine λ: `:fixed`, `:gcv_svd`, `:gcv_tr`, `L_curve`
+    Aitp::AbstractInterpolation{FT,T}
+    RegularizationSmooth{FT}(u,û,t,t̂,d,λ,alg,Aitp) where FT =
+        new{typeof(u),typeof(t),FT,eltype(u)}(u,û,t,t̂,d,λ,alg,Aitp)
+end
+
+function RegularizationSmooth(u, t, d=2; λ=1.0, t̂=nothing, alg=:gcv_svd)
+    # TBD:  make use of `munge_data`
+    u, t = munge_data(u, t)
+    λ = float(λ) # in case an integer is given for λ
+    N = length(t)
+    if isnothing(t̂)
+        t̂ = t
+        # identity matrix for A (RegularizationTools `A` matrix, not `A` object of
+        # DataInterpolations)
+        A = Array{Float64}(LA.I, (N, N))
```

Comment:
this should probably be eltype matching

Reply:
Got it. Also now using eltype in `_mapping_matrix`.